### PR TITLE
fix(helper): don't let the accept deadline bound the whole transfer (#44)

### LINKED ISSUE #44: Large-table copy aborts with SQLSTATE 22P04 after exactly 600s (ExtDestLtCopy + push): listener deadline caps total transfer time
## Summary

When copying a table whose data transfer takes longer than **600 seconds**, cbcopy aborts it mid-stream: the destination fails with `ERROR: missing data for column "..." (SQLSTATE 22P04)` and the source with `could not write to COPY program: Broken pipe`. The failure lands at almost exactly the 10-minute mark and is independent of network health (it reproduces even when full-mesh segment-to-segment connectivity is verified healthy).

The root cause is a fixed 600 s deadline on the receive-side helper's **listener** that, for one specific copy strategy, ends up capping the **total transfer time** of a table instead of just the connection-accept phase.

Reproduces on current `main` (`helper/server.go:82`). Also reported (v1.1.4) at apache/cloudberry#1830.

## Root cause

`ServerBase.Start()` sets an absolute deadline on the TCP listener:

```go
// helper/server.go:82
err = tListener.SetDeadline(time.Now().Add(time.Second * 600))
```

For a `ConcurrentServer`, `serve()` runs an unbounded `Accept()` loop (`helper/server.go:130`), and `handleConnection` hands each connection to a goroutine and returns immediately (`helper/concurrent_server.go:34-36`). So once all expected clients (`--client-numbers`) have connected, the loop calls `Accept()` again and blocks there **for the entire lifetime of the transfer**. When the absolute 600 s deadline expires, that still-blocked `Accept()` returns `i/o timeout`, which calls `setError()`; `WaitForFinished()` then returns because `err != nil` (`helper/server.go:111`), the helper exits with an error, and the in-flight data connections are torn down. The destination QE reads a truncated row → 22P04; the source sees broken pipe / connection reset.

In other words, a deadline that (I assume) was intended to bound how long we wait for inbound connections also bounds how long the transfer itself may run.

## When it triggers — and when it doesn't

`--client-numbers` is the only flag that makes `createServerImpl` build a `ConcurrentServer` (`helper/server.go:160-167`, when a value is `> 1`). It is emitted at exactly one place in the whole codebase:

- `copy/copy_command.go:449` — `ExtDestLtCopy.CopyFrom`, **push** branch (destination cluster has fewer segments than source, so each destination segment listens for multiple source segments).

Every other `--listen` site (`CopyOnMaster`, `CopyOnSegment`, `ExtDestGeCopy`, and all pull-mode senders) omits `--client-numbers`, so those run as `OneTimeServer`, which closes its listener immediately after the single `Accept()` (`helper/one_time_server.go:28-34`) and then streams with no deadline — unaffected.

So the bug is confined to **`ExtDestLtCopy` + push mode** (destination has fewer segments than source; `copy_command.go:596`), on destination segments assigned more than one source segment, and only for tables whose transfer exceeds 600 s. That precisely matches the observed pattern: *small tables succeed, large tables fail, always at ~10 minutes*.

`--connection-mode=pull` sidesteps it (the source side listens → `OneTimeServer`), which is a usable workaround when dest < src.

## Reproduction

1. A source cluster with more segments than the destination (so `ExtDestLtCopy` is selected).
2. Copy a single large table in the default push mode, e.g. `cbcopy ... --include-table <big.table> --data-only --truncate`.
3. Any table whose data takes longer than ~10 minutes to transfer aborts with SQLSTATE 22P04 at ~600 s; the source log shows `connection reset by peer` / `Broken pipe` and the destination helper log shows a flood of `Listener.Accept() failed: accept tcp [::]:<port>: i/o timeout` at exactly start + 600 s.

## Suggested fix

The `ConcurrentServer` should not let the listener deadline bound the transfer. Two options:

1. Once `numConn` connections have been accepted, stop calling `Accept()` (or clear the deadline via `SetDeadline(time.Time{})`) so the transfer runs without a listener timeout; or
2. Replace the fixed listener deadline with an **idle / no-progress timeout on the data connections** — reset a read deadline whenever bytes arrive. That keeps the "don't hang forever on a genuinely stuck connection" protection while never tearing down a transfer that is still making progress.

I'm happy to send a PR — I'd just like to confirm the intended semantics of the 600 s value first (accept-phase bound vs. overall guard).

### PR DESCRIPTION
**Pull request by talmacschen-arch**
fix #44

---

## Change logs

`ConcurrentServer.serve()` runs an unbounded `Accept()` loop that lives for the whole transfer. The listener's absolute 600 s deadline (set in `ServerBase.Start()`) therefore bounds the **total transfer time** of a single table, not just the accept phase: once all expected clients (`--client-numbers`) have connected, the loop blocks on `Accept()` again, and at start + 600 s that still-blocked `Accept()` returns `i/o timeout` → `setError` → the transfer is torn down. This is confined to `ExtDestLtCopy` + push (destination has fewer segments than source), i.e. any table whose data takes longer than ~10 minutes to move. See #44 for the full root-cause analysis.

**Fix:** make `serve()` accept-count aware — stop accepting once `numConn` connections are in, then `Close()` the listener so the in-flight data connections stream to completion with no listener deadline over them. A timeout/error *before* all expected clients have connected is still fatal (a client genuinely failed to connect), and the error path now `return`s instead of `continue`-ing, so a tripped deadline can no longer busy-spin flooding the log.

**Scope:** only `ConcurrentServer` (`ExtDestLtCopy` + push) is affected. `OneTimeServer` (`CopyOnMaster` / `CopyOnSegment` / `ExtDestGeCopy` and all pull-mode senders) closes its listener after a single `Accept()` and is unchanged — so `--connection-mode=pull` was, and remains, unaffected.

## Validation

On a 16 → 8 segment setup (source 16 primaries, destination 8) copying a single large unpartitioned table (TPC-DS `store_sales`, 1 TB scale) in default push mode:

- The pre-fix helper reproduces #44 at exactly start + 600 s — as an abort (`could not write to COPY program: Broken pipe` on the source; `22P04` / `25M01` on the destination) and, depending on timing, as an indefinite hang with no error surfaced to either side.
- After rebuilding just the receive-side helper with this change and re-running the same copy, it ran well past the 600 s cap and completed normally (`successfully copied 1 tables`, total elapsed 36m52s), with **zero** `Listener.Accept() ... i/o timeout` on the destination helpers.

The `helper` package currently has no unit tests; happy to add a focused `serve()` test (accept `numConn` → listener closed, `Err()==nil`, no timeout; fewer than `numConn` before the deadline → `Err()!=nil`, no busy-spin) if maintainers would like one in this PR.

## Summary by CodeRabbit

* **Bug Fixes**
  * Server listeners now stop accepting connections after the expected number of clients connect.
  * Accept errors now terminate the connection-acceptance process promptly instead of being ignored.
  * Listeners close automatically once all expected connections are established, while existing connections remain active.

### CODE REVIEW

File: helper/server.go [hunk 1]
```diff
@@ -130,21 +130,33 @@ func (t *ServerBase) Serve() {
...
 		}
 
+		accepted++
 		t.handler.handleConnection(conn)
 	}
+
+	// All expected clients are connected. Stop letting the listener's accept
+	// deadline bound the rest of the transfer: close the listener so the
+	// in-flight data connections stream to completion with no timeout (issue #44).
+	_ = t.listener.Close()
```
[the remark this answers is not in the document]
Comment (talmacschen-arch) [reply helper/server.go]: Thanks — surfacing the close failure is reasonable, done in d5f4e86.

I intentionally kept it **log-only** rather than mirroring `one_time_server.go`'s `setError` on close, because the semantics differ. `one_time_server.go` closes its listener *before* streaming, so a close failure there is relevant to its flow. Here `serve()` closes the listener *after* all expected clients (`numConn`) have connected and the data goroutines are already streaming — at that point a listener-close failure is benign and doesn't affect the transfer. Calling `setError` would make `WaitForFinished()` return with `err != nil` and tear down an otherwise-healthy transfer, which is exactly the spurious-teardown class of bug this PR (#44) is fixing. So I log at debug for visibility but don't fail the transfer.

**Review by leaocx — APPROVED** [review]

**Review by leaocx — APPROVED** [review]
LGTM. Precise fix with solid validation.

- Bounded accept loop correctly prevents the listener deadline from capping the entire transfer
- `continue` → `return` on error path eliminates busy-spin
- Closing the listener does not affect already-established data connections
- Log-only (no `setError`) on listener-close failure is the right call — avoids reintroducing spurious teardown
